SAT/Z3Interfacing.cpp

```diff
@@ -98,50 +98,63 @@ const char* errToString(Z3_error_code code)
   }
 }
 
-void handleZ3Error(Z3_context ctxt, Z3_error_code code) 
+void handleZ3Error(Z3_context ctxt, Z3_error_code code)
 {
   DEBUG(errToString(code))
   throw z3::exception(errToString(code));
 }
 
 #define STATEMENTS_TO_EXPRESSION(...) [&]() { __VA_ARGS__; return 0; }()
 
-Z3Interfacing::Z3Interfacing(SAT2FO& s2f, bool showZ3, bool unsatCore):
-  _varCnt(0), 
+Z3Interfacing::Z3Interfacing(SAT2FO& s2f, bool showZ3, bool unsatCore, vstring const& exportSmtlib):
+  _varCnt(0),
   _sat2fo(s2f),
-  _status(SATISFIABLE), 
+  _status(SATISFIABLE),
   _config(),
   _context(_config),
   _solver(_context),
   _model((STATEMENTS_TO_EXPRESSION(
-            BYPASSING_ALLOCATOR; 
-            _solver.check(); 
+            BYPASSING_ALLOCATOR;
+            _solver.check();
           ),
-         _solver.get_model())), 
+         _solver.get_model())),
   _assumptions(),
   _showZ3(showZ3),
-  _unsatCore(unsatCore)
+  _unsatCore(unsatCore),
+  _out(exportSmtlib == "" ? Option<std::ofstream>()
+                          : Option<std::ofstream>(std::ofstream(exportSmtlib.c_str())) )
 {
   CALL("Z3Interfacing::Z3Interfacing");
   BYPASSING_ALLOCATOR
   _solver.reset();
 
-  z3::set_param("rewriter.expand_store_eq", "true");
+  if (_out.isSome() && _out.unwrap().fail()) {
+    throw UserErrorException("Failed to open file: ", exportSmtlib);
+  }
+  outputln("(check-sat)");
+  outputln("(get-model)");
+  outputln("(reset)");
 
   z3::params p(_context);
-  p.set("model.compact", false); // keeps z3 from compressing its model. ~50% of the runtime of get_model is spent doing that otherwise
+  auto setOption = [&](auto k, auto v){
+    p.set(k,v);
+    outputln(";- z3 parameter: ", k, "=", v);
+  };
+  setOption("rewriter.expand_store_eq", true);
```

Comment:
Can you add a comment about what this option is doing?

Reply:
Tbh I don't have any idea. the option was present before. I just kept it. But I'm happy to remove it as it would simplify the code ;) 
Git blame says @quickbeam123 has added the option in the first place.
`z3 -pd` explains
```
    expand_store_eq (bool) reduce (store ...) = (store ...) with a common base into selects (default: false)
```
I reckon we want to keep that.. (?)

Reply:
Sorry, I assumed it was new as I didn't see it removed from elsewhere. I  can see in another thread that this option was suggested by Nikolaj so let's definitely keep it ;)